src/openrct2/core/FileScanner.cpp

```diff
@@ -50,72 +50,62 @@ struct DirectoryChild
     uint64_t LastModified = 0;
 };
 
-static uint32_t GetPathChecksum(const utf8* path);
+static uint32_t GetPathChecksum(u8string_view path);
 static bool MatchWildcard(const utf8* fileName, const utf8* pattern);
 
 class FileScannerBase : public IFileScanner
 {
 private:
     struct DirectoryState
     {
-        std::string Path;
+        u8string Path;
         std::vector<DirectoryChild> Listing;
         int32_t Index = 0;
     };
 
     // Options
-    std::string _rootPath;
-    std::vector<std::string> _patterns;
-    bool _recurse;
+    const u8string _rootPath;
+    const std::vector<std::string> _patterns;
+    const bool _recurse;
 
     // State
     bool _started = false;
     std::stack<DirectoryState> _directoryStack;
 
     // Current
-    FileScanner::FileInfo* _currentFileInfo;
-    utf8* _currentPath;
+    FileScanner::FileInfo _currentFileInfo;
+    u8string _currentPath;
 
 public:
-    FileScannerBase(const std::string& pattern, bool recurse)
+    FileScannerBase(u8string_view pattern, bool recurse)
+        : _rootPath(Path::GetDirectory(pattern))
+        , _patterns(GetPatterns(Path::GetFileName(pattern)))
+        , _recurse(recurse)
     {
-        _rootPath = Path::GetDirectory(pattern);
-        _recurse = recurse;
-        _patterns = GetPatterns(Path::GetFileName(pattern));
-
-        _currentPath = Memory::Allocate<utf8>(MAX_PATH);
-        _currentFileInfo = Memory::Allocate<FileScanner::FileInfo>();
-
-        Reset();
     }
 
-    ~FileScannerBase() override
-    {
-        Memory::Free(_currentPath);
-        Memory::Free(_currentFileInfo);
-    }
+    ~FileScannerBase() override = default;
 
-    const FileScanner::FileInfo* GetFileInfo() const override
+    const FileScanner::FileInfo& GetFileInfo() const override
     {
         return _currentFileInfo;
     }
 
-    const utf8* GetPath() const override
+    const u8string& GetPath() const override
     {
         return _currentPath;
     }
 
-    const utf8* GetPathRelative() const override
+    u8string GetPathRelative() const override
     {
-        // +1 to remove the path separator
-        return _currentPath + _rootPath.size() + 1;
+        return Path::GetRelative(_currentPath, _rootPath);
     }
 
     void Reset() override
     {
         _started = false;
-        _directoryStack = std::stack<DirectoryState>();
-        _currentPath[0] = 0;
+        std::stack<DirectoryState>().swap(_directoryStack);
```

Comment:
```suggestion
        _directoryStack = {};
```
That should be sufficient to reset it